module/zfs/zfs_ioctl.c

```diff
@@ -3561,8 +3561,40 @@ zfs_ioc_destroy(zfs_cmd_t *zc)
 
 	if (strchr(zc->zc_name, '@'))
 		err = dsl_destroy_snapshot(zc->zc_name, zc->zc_defer_destroy);
-	else
+	else {
 		err = dsl_destroy_head(zc->zc_name);
+		if (err == EEXIST) {
+			/*
+			 * It is possible that the given DS may have
+			 * hidden child (%recv) datasets - "leftovers"
+			 * resulting from the previously interrupted
+			 * 'zfs receive'.
+			 */
+			char namebuf[MAXNAMELEN];
+
+			(void) snprintf(namebuf, sizeof (namebuf),
+			    "%s/%%recv", zc->zc_name);
+
+			/* Try to remove the hidden child (%recv) */
+			err = dsl_destroy_head(namebuf);
+			if (err == 0) {
+				/*
+				 * Now the given DS should not have
+				 * children, so we can try to remove
+				 * it again
+				 */
```

Comment:
I'd suggest dropping this comment and the next one since I don't think they add anything.  There code here is very clear already.  You might consider consolidating the reasoning here in to the `Try to remove...` comment above.